Update Password Form in UserService to KS Components (ILIAS 10)

fhelfer opened the pull request:
Updated the User Password Form in UserService with new KS UI Components.
After consultation with @kergomard, we simply use one input for the new password. No need to duplicate this if the plain text display is available via the eye glyph.

There is also the question @kergomard, whether you still want the code to be moved to Auth for this PR?

Best, @fhelfer

fhelfer committed:
Update User Password Change Form to KS Components

kergomard commented:
Hi @fhelfer 

Thank you very much for the PR and your question. I wait for you to give me the green light before looking at this, but I can at least answer your question:
My feeling is, that the password does not belong to `ilUser` and thus it should be move to `Auth`, but this is in no way a requirement. At the end, I think it is the password as well as the login (unless we just specify it as a nickname) that should be moved. There is also `ilUserPasswordManager` and `ilUserPasswordEncoderFactory` that should go with it and all places where passwords (and logins) can be changed. So if you believe I'm wrong or if you believe that we should do this in a more coordinated fashion and need more time to do so, you can also keep it where it is (for now). Please just argue in a few sentences why it should be so.

Thanks again and best,
@kergomard

mjansenDatabay commented:
@kergomard We will move the `ilUserPassword*` classes to `Services/Auth` as well as the password-related methods in `ilPersonalSettingsGUI` (will be moved to a new GUI class).

fhelfer committed:
Move User Password Settings to Auth

fhelfer committed:
Fix Password Test

Fix Password Test 2

fhelfer commented:
Hi @kergomard 
The work is done
Files have been moved to Authentication
PR is no longer WIP

Best, @fhelfer

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+                    $this->tpl->setOnScreenMessage('success', $this->lng->txt('saved_successfully'), true);
+                    $target = ilSession::get('orig_request_target');
+                    ilSession::set('orig_request_target', '');
+                    ilUtil::redirect($target);
```
Can we use `redirectToUrl` from `ilCtrl` instead?

kergomard commented:
Thank you very much @fhelfer!

I only looked at the code that stayed in `ilUser` and I have only three very little things:
- Shouldn't we also move the [`ilObjPasswortTest`](https://github.com/ILIAS-eLearning/ILIAS/pull/7370/files#diff-6cc9498e5da3028fddc28904175786540da0c174e5dec24e26038ff19cb32b35)?
- Please move the logical operators back to the beginning of the line where you have moved them to the end, e.g. [here](https://github.com/ILIAS-eLearning/ILIAS/pull/7370/files#diff-5ab2ee5bf5242a5b6aacd882774a51562e18a687f20a0049457ff9f1f1f5b8b5R155).
- Please add the second asterisk again to the @var comment [here](https://github.com/ILIAS-eLearning/ILIAS/pull/7370/files#diff-5ab2ee5bf5242a5b6aacd882774a51562e18a687f20a0049457ff9f1f1f5b8b5R56). One as it is a DOC-Block it is needed, I think, but two and more important for me: If it is not there Netbeans does not recognize the variable.

Thanks and best,
@kergomard

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+    private ilLanguage $lng;
+    private ilObjUser $user;
+    private ilCtrlInterface $ctrl;
+    private ilLocalPasswordManager $password_manager;
```
Can these be made `readonly`?

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+        $cmd = $this->ctrl->getCmd();
+        switch ($cmd) {
+            default:
+                $this->$cmd();
```
We should check whether the command exists (as a method) and make the code robust (throw a `failure` if the command method does not exist).

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+        // check whether password of user have to be changed
+        // due to first login or password of user is expired
+        if ($this->user->isPasswordChangeDemanded()) {
+            $this->tpl->setOnScreenMessage('info', $this->lng->txt('password_change_on_first_login_demand'));
```
Could we use a `UI/Mesage > Info` instead`?

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+        } elseif ($this->user->isPasswordExpired()) {
+            $msg = $this->lng->txt('password_expired');
+            $password_age = $this->user->getPasswordAge();
+            $this->tpl->setOnScreenMessage('info', sprintf($msg, $password_age));
```
Could we use a `UI/Mesage > Info` instead`?

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+                    $cpass = $cpass->withError(implode('<br>', $cpass_error));
+                }
+
+                $items['current_password'] = $cpass;
```
Could we add `withAdditionalTransformation` for the validations you currently do in `savePassword`?

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+            if ($ipass_error !== []) {
+                $ipass = $ipass->withError(implode('<br>', $ipass_error));
+            }
+            $items['new_password'] = $ipass;
```
Could we add `withAdditionalTransformation` for the validations you currently do in `savePassword`?

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+
+        if (!$form->getError()) {
+            $data = $form->getData();
+            $error = false;
+
+            if ((int) $this->user->getAuthMode(true) === ilAuthUtils::AUTH_LOCAL) {
+                if (!ilLocalPasswordManager::getInstance()->verifyPassword(
+                    $this->user,
+                    $entered_current_password
+                )) {
+                    $error = true;
+                    $errors['current_password'][] = $this->lng->txt('passwd_wrong');
+                }
+            }
+
+            if (!ilSecuritySettingsChecker::isPassword($entered_new_password, $custom_error)) {
+                $error = true;
+                if ($custom_error !== '') {
+                    $errors['new_password'][] = $custom_error;
+                } else {
+                    $errors['new_password'][] = $this->lng->txt('passwd_invalid');
+                }
+            }
+            $error_lng_var = '';
+            if (!ilSecuritySettingsChecker::isPasswordValidForUserContext(
+                $entered_new_password,
+                $this->user,
+                $error_lng_var
+            )) {
+                $this->tpl->setOnScreenMessage('failure', $this->lng->txt('form_input_not_valid'));
+                $errors['new_password'][] = $this->lng->txt($error_lng_var);
+                $error = true;
+            }
+            if (
+                $entered_current_password === $entered_new_password &&
+                ($this->user->isPasswordExpired() || $this->user->isPasswordChangeDemanded())
+            ) {
+                $error = true;
+                $errors['new_password'][] = $this->lng->txt('new_pass_equals_old_pass');
+            }
```
See my comment above, please check whether some code lines could be moved to transformations of the defined inputs.

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+                }
+
+                if (ilSession::get('orig_request_target')) {
+                    $this->tpl->setOnScreenMessage('success', $this->lng->txt('saved_successfully'), true);
```
```suggestion
                    $this->tpl->setOnScreenMessage($this->tpl::MESSAGE_TYPE_SUCCESS, $this->lng->txt('saved_successfully'), true);
```

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.ilLocalPasswordSettingsGUI.php:
```diff
@@ -0,0 +1,249 @@
+                    ilSession::set('orig_request_target', '');
+                    ilUtil::redirect($target);
+                } else {
+                    $this->tpl->setOnScreenMessage('success', $this->lng->txt('saved_successfully'));
```
```suggestion
                    $this->tpl->setOnScreenMessage($this->tpl::MESSAGE_TYPE_SUCCESS, $this->lng->txt('saved_successfully'));
```

mjansenDatabay replied:
Or even better: Can this re replaced with a `UI/Message > Success` element and (because no HTTP redirect is done) passed to `showPassword` (where it should be rendered above the form)?

mjansenDatabay reviewed (changes requested):
Thank you very much @fhelfer ,

besides my inline comments I have the following remarks:

- I suggest to rename the `ilLocalPassword*` files to `ilLocalUserPassword*`.
- As @kergomard already mentioned, the test class should be moved and renamed as well.
- You introduced some code style chances. Why :-)?

Best regards,
Michael

fhelfer committed:
Contraints on LocalUserPasswordSettingsForm

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordSettingsGUI.php:
```diff
@@ -65,30 +71,35 @@ public function executeCommand(): void
+                if (method_exists($this, $cmd)) {
+                    $this->$cmd();
+                } else {
+                    throw new ilException('');
```
Please use:

```php
// ...
$this->error = $DIC['ilErr'];
// ...
$this->error->raiseError($this->lng->txt('permission_denied'), $this->error->MESSAGE);
// ...
```

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordSettingsGUI.php:
```diff
@@ -114,6 +125,16 @@ public function getPasswordForm(
                     $cpass = $cpass->withError(implode('<br>', $cpass_error));
                 }
+                $cpass = $cpass->withAdditionalTransformation(
+                    $this->refinery->custom()->constraint(function (Password $value) {
```
```suggestion
                    $this->refinery->custom()->constraint(function (Password $value): bool {
```

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordSettingsGUI.php:
```diff
@@ -131,6 +152,40 @@ public function getPasswordForm(
                 $ipass = $ipass->withError(implode('<br>', $ipass_error));
             }
+            $ipass = $ipass->withAdditionalTransformation(
+                $this->refinery->custom()->constraint(function (Password $value) {
```
```suggestion
                $this->refinery->custom()->constraint(function (Password $value): bool {
```

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordSettingsGUI.php:
```diff
@@ -131,6 +152,40 @@ public function getPasswordForm(
+            $ipass = $ipass->withAdditionalTransformation(
+                $this->refinery->custom()->constraint(function (Password $value) {
+                    return ilSecuritySettingsChecker::isPassword($value->toString(), $custom_error);
+                }, function (Closure $txt, Password $value) {
```
```suggestion
                }, function (Closure $txt, Password $value): string {
```

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordSettingsGUI.php:
```diff
@@ -131,6 +152,40 @@ public function getPasswordForm(
+            );
+            $ipass = $ipass->withAdditionalTransformation(
+                $this->refinery->custom()->constraint(
+                    function (Password $value) {
```
```suggestion
                    function (Password $value): bool {
```

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordSettingsGUI.php:
```diff
@@ -131,6 +152,40 @@ public function getPasswordForm(
+                            $error_lng_var
+                        );
+                    },
+                    function (Closure $cls, Password $value) {
```
```suggestion
                    function (Closure $cls, Password $value): string {
```

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordSettingsGUI.php:
```diff
@@ -172,48 +227,28 @@ public function savePassword(): void
         $section = $form->getInputs()['password'];
-        /** @var Password $cp
-         * @var Password $np
+        /** @var PasswordInput $cp
+         * @var PasswordInput $np
```
```suggestion
        /** 
         * @var PasswordInput $cp
         * @var PasswordInput $np
```

mjansenDatabay commented on components/ILIAS/Authentication/tests/ilLocalUserPasswordTest.php:
```diff
@@ -18,17 +18,18 @@
 require_once __DIR__ . '/ilUserBaseTest.php';
 
-class ilObjUserPasswordTest extends ilUserBaseTest
+class ilLocalUserPasswordTest extends ilUserBaseTest
```
You have to register/include the test class in `ilServicesAuthenticationSuite`

fhelfer committed:
Code Style LocaUserPasswordAuth

fhelfer committed:
remove operator_linebreak cs fixer changes

fhelfer commented:
> * You introduced some code style chances. Why :-)?

I have a CS-fixer Configuration with more rules than ILIAS (mostly personal preferences, most of which are PSR12 compatible and just add not configured code styles, e.g. Symfony ruleset)
Since I moved the whole file, I thought might as well make it look pretty :shrug: 

> lease move the logical operators back to the beginning of the line where you have moved them to the end, e.g. [here](https://github.com/ILIAS-eLearning/ILIAS/pull/7370/files#diff-5ab2ee5bf5242a5b6aacd882774a51562e18a687f20a0049457ff9f1f1f5b8b5R155).

@kergomard I think I removed everywhere I changed this. Its due to
`'operator_linebreak' => ['only_booleans' => true, 'position' => 'end'],`
Its 'beginning as default in Symfony Ruleset. I have seen both mixed in ILIAS so maybe adding one way to code_format might be good? :)
Also, [phpdoc_to_comment](https://cs.symfony.com/doc/rules/phpdoc/phpdoc_to_comment.html) makes comments single quote in Symfony Ruleset

Best, @fhelfer

kergomard reviewed (approved):
Thanks @fhelfer !
I'm fine with this now. I'm not sure we want to introduce more rules here, but I'm also not going to die on this hill. I believe that operators at the beginning have a rather important advantage in readability and that is why I prefer them there.

Thanks again and best,
@kergomard

fhelfer committed:
Fix LocalUserPasswordTest not found

revert reformatting

revert reformatting

constants new/current_password

mjansenDatabay commented:
@pascalseeland Any objections here?

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordManager.php:
```diff
@@ -18,21 +18,31 @@
+class LocalUserPasswordManager
 {
     private const MIN_SALT_SIZE = 16;
-
```
Please re-introduce the empty lines as an optical separator between constants, static properties and members.

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordEncoderFactory.php:
```diff
@@ -129,7 +140,8 @@ public function getEncoderByName(?string $name): ilPasswordEncoder
             throw new ilUserException('No default encoder specified, fallback not possible.');
-        } elseif (!isset($this->supported_encoders[$this->getDefaultEncoder()])) {
+        }
+        if (!isset($this->supported_encoders[$this->getDefaultEncoder()])) {
```
```suggestion

        if (!isset($this->supported_encoders[$this->getDefaultEncoder()])) {
```

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordManager.php:
```diff
@@ -42,32 +52,37 @@ public function __construct(array $config = [])
-                '"encoder_factory" must be instance of ilUserPasswordEncoderFactory and set in %s.',
-                print_r($config, true)
-            ));
+        if (!$this->getEncoderFactory() instanceof LocalUserPasswordEncoderFactory) {
+            throw new ilUserException(
+                sprintf(
+                    '"encoder_factory" must be instance of ilUserPasswordEncoderFactory and set in %s.',
+                    print_r($config, true)
+                )
+            );
```
Please revert the reformatting.

pascalseeland replied:
And also please fix the message to state LocalUserPasswordEncoder instead on ilUserPass..

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordManager.php:
```diff
@@ -42,32 +52,37 @@ public function __construct(array $config = [])
-                '"password_encoder" must be set in %s.',
-                print_r($config, true)
-            ));
+            throw new ilUserException(
+                sprintf('"password_encoder" must be set in %s.', print_r($config, true))
+            );
```
Please revert the reformatting.

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordManager.php:
```diff
@@ -84,23 +99,27 @@ public static function getInstance(): self
-                        'default_password_encoder' => 'bcryptphp', // bcrypt (native PHP impl.) is still the default for the factory
-                        'memory_cost' => 19_456, // Recommended: https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
+                        'default_password_encoder' => 'bcryptphp',
+                        // bcrypt (native PHP impl.) is still the default for the factory
```
Please move this line up again or put it on top of the line before.

mjansenDatabay commented on components/ILIAS/Authentication/classes/Password/class.LocalUserPasswordManager.php:
```diff
@@ -84,23 +99,27 @@ public static function getInstance(): self
                 ),
-                'password_encoder' => 'argon2id', // https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
+                'password_encoder' => 'argon2id',
+                // https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id
```
Please move this line up again or put it on top of the line before.
